src/gurt/telemetry.c

```diff
@@ -834,12 +780,109 @@ d_tm_compute_duration_stats(struct d_tm_node_t *node)
  * faster access.
  *
  * \param[in,out]	metric	Pointer to the metric
- * \param[in]		item	Full path name to this \a item
- *				If supplied, the path name is specified by the
- *				\a item and any strings following it.  The last
- *				item must be NULL.
+ * \param[in]		fmt	Format specifier for the name and full path of
+ *				the metric followed by optional args to
+ *				populate the string, printf style.
+ *				The format specifier and optional arguments
+ *				are used only if the pointer to the metric
+ *				is NULL.
+ *
+ * \return			DER_SUCCESS		Success
+ *				-DER_EXCEEDS_PATH_LEN	The full name length is
+ *							too long
+ *				-DER_OP_NOT_PERMITTED	Operation not permitted
+ *							because the \a item is
+ *							not a counter
+ *				-DER_UNINIT		API not initialized
+ *				-DER_INVAL		\a metric and \a item
+ *							are NULL
+ */
+int
+d_tm_set_counter(struct d_tm_node_t **metric, uint64_t val,
+		 const char *fmt, ...)
+{
+	struct d_tm_node_t	*node = NULL;
+	char			path[D_TM_MAX_NAME_LEN] = {};
+	int			rc = DER_SUCCESS;
+
+	if (d_tm_shmem_root == NULL)
+		return -DER_UNINIT;
+
+	if ((metric != NULL) && (*metric != NULL)) {
+		node = *metric;
+	} else {
+		va_list	args;
+		int	ret;
+
+		if (fmt == NULL) {
+			rc = -DER_INVAL;
+			goto failure;
+		}
+
+		va_start(args, fmt);
+		ret = vsnprintf(path, sizeof(path), fmt, args);
+		va_end(args);
+
+		if (!((ret > 0) && (ret < D_TM_MAX_NAME_LEN))) {
+			rc = -DER_EXCEEDS_PATH_LEN;
+			goto failure;
+		}
+
+		node = d_tm_find_metric(d_tm_shmem_root, path);
+		if (metric != NULL)
+			*metric = node;
+	}
+
+	if (node == NULL) {
+		rc = d_tm_add_metric(&node, D_TM_COUNTER, "N/A", "N/A", path);
+		if (rc != DER_SUCCESS) {
+			D_ERROR("Failed to add and incremement counter [%s]: "
+				DF_RC "\n", path, DP_RC(rc));
+			goto failure;
+		}
+		if (metric != NULL)
+			*metric = node;
+	}
+
+	if (node->dtn_type == D_TM_COUNTER) {
+		if (node->dtn_protect)
+			D_MUTEX_LOCK(&node->dtn_lock);
+		node->dtn_metric->dtm_data.value = val;
+		if (node->dtn_protect)
+			D_MUTEX_UNLOCK(&node->dtn_lock);
+	} else {
+		rc = -DER_OP_NOT_PERMITTED;
+		D_ERROR("Failed to set counter [%s] on item not a "
+			"counter.  Operation mismatch: " DF_RC "\n",
+			node->dtn_name, DP_RC(rc));
+		goto failure;
+	}
```

Comment:
minor style nit: It would probably be clearer to invert these clauses by checking `node->dtn_type != D_TM_COUNTER` and then you could get rid of the else clause by outdenting the "happy path" code.

Reply:
Sure, I'll make that change to these producer functions.